### ISSUE TITLE
Broken compilation of size-solver

### ISSUE THREAD

**Issue by asr** [agda/agda#1939]
While working on #1828, I found the following error:

```
$ cd src/size-solver
$ cabal install
Parser.hs:93:17:
    Couldn't match type ‘Either String [Constraint' Rigid Flex]’
                   with ‘Maybe [Constraint]’
    Expected type: Constraint -> Maybe [Constraint]
      Actual type: CTrans Rigid Flex
    In the first argument of ‘map’, namely
      ‘(simplify1 (\ c -> Just [c]))’
    In the expression: map (simplify1 (\ c -> Just [c]))
```

After fixing this issue and #1828, I will fix #1822.

**Comment by asr** [comment]
The error is due to changes introduced in 02846e4.

**Comment by andreasabel** [comment]
Ah sorry, I did not try to compile the size-solver.  Some functions now return an error message instead of just "Nothing".  Let me know if I should take over (currently you are assigned to this issue).

**Comment by asr** [comment]
I'll push a fix in a few minutes.

**Comment by asr** [comment]
I fixed this issue in 09d20bb (but I wrote wrong the commit message).

**Comment by andreasabel** [comment]
I am sorry I caused WWII.  (As always, it's the German's fault.) :D

**Comment by asr** [comment]
I back-ported the fix and the refactoring into maint-2.4.
